crates/kitsune_p2p/mdns/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+///! Crate for discovering Holochain peers over MDNS
+///! Works by broadcasting a service named `HC_SERVICE_NAME`
+///! and adding base64 encoded data in a TXT record
+///!
+///! Uses libmdns crate for broadcasting
+///! Uses mdns crate for discovery
+
+use futures_util::{stream::StreamExt, self};
+use std::thread;
+use std::time::Duration;
+use std::sync::mpsc::{self, TryRecvError, Sender};
+use mdns::RecordKind;
+use futures_core::Stream;
+use err_derive::Error;
+
+const HC_SERVICE_NAME: &str       = "_holochain._udp";
+const BROADCAST_INTERVAL_SEC: u64 = 8;
+const QUERY_INTERVAL_SEC: u64     = 5;
+
+#[derive(Debug, Error)]
+pub enum MdnsError {
+   #[error(display = "Regular Mdns error {}", _0)]
+   Mdns(#[error(source)] mdns::Error),
+   #[error(display = "Base64 decoding error {}", _0)]
+   Base64(#[error(source)] base64::DecodeError),
+}
+
+/// Stop thread created by `mdns_create_broadcast_thread()`
+pub fn mdns_kill_thread(tx: Sender<()>) {
+   tx.send(()).ok();
+}
+
+/// Create a thread that will broadcast the holochain service over mdns
+/// Returns Sender for sending thread termination command
+pub fn mdns_create_broadcast_thread(buffer: &[u8]) -> Sender<()> {
+   // Create Terminate command channel
+   let (tx, rx) = mpsc::channel();
+   // Change buffer to base64 string
+   let b64 =format!("u{}", base64::encode_config(buffer, base64::URL_SAFE_NO_PAD));
+   println!("base64 length is {}", b64.len());
+
+   // Create thread
+   let _handle = thread::spawn(move || {
+      // debug
+      println!("Entering mdns broadcasting thread...");
+      // Create mdns responder
+      let responder = libmdns::Responder::new().unwrap();
+      let _svc = responder.register(
+         HC_SERVICE_NAME.to_owned(),
+         "holonode".to_owned(),
```

Comment:
I wonder if there is utility in setting this to the AgentKey?

Reply:
Actually, now that I think about it... We're going to need to publish one of these for each DnaHash/AgentKey combo - so we probably need to set this to that so they can be distinguished.

Is it possible for one machine to publish multiple responders for the same HC_SERVICE_NAME like that?

Reply:
Yes, at least I tried two on my windows machine. So yes instead of "holonode" we can set the dnaHash.
I'm renaming `HC_SERVICE_NAME` to `HC_SERVICE_TYPE` to avoid confusion

Reply:
Oh wait, the name can't be longer than 63 chars, so a dnaHash might not fit.

Reply:
That 63 limit seems arbitrary, we might be able to just remove the assert on that.

Reply:
the string repr should be 53 characters long.
but, i think we need the agent key too. because you can have multiple agents using the same dna on the same conductor.

Reply:
unless we want to make the service type be the dna hash... because in some ways, we don't care about broadcasts for DNAs that we haven't joined

Reply:
service type has the same 63 limit

Reply:
Right, so we can set the service type to the 53 characters of the dnaHash, and set the name to the 53 charaters of the agentKey.